eth/tracers/native/contract.go

```diff
@@ -0,0 +1,134 @@
+	"github.com/XinFinOrg/XDPoSChain/core/vm"
+	"github.com/XinFinOrg/XDPoSChain/eth/tracers"
+)
+
+func init() {
+	tracers.RegisterNativeTracer("contractTracer", NewContractTracer)
+}
+
+type contractTracer struct {
+	Addrs     map[string]string
+	config    contractTracerConfig
+	interrupt uint32 // Atomic flag to signal execution interruption
+	reason    error  // Textual reason for the interruption
+}
+
+type contractTracerConfig struct {
+	OpCode       string `json:"opCode"`       // Target opcode to trace
+	WithByteCode bool   `json:"withByteCode"` // If true, bytecode will be collected
+}
+
+// NewContractTracer returns a native go tracer which tracks the contracr was created
+func NewContractTracer(cfg json.RawMessage) (tracers.Tracer, error) {
+	var config contractTracerConfig
+	if cfg != nil {
+		if err := json.Unmarshal(cfg, &config); err != nil {
+			return nil, err
+		}
+	}
+	t := &contractTracer{
+		Addrs:  make(map[string]string, 1),
+		config: config,
+	}
+	// handle invalid opcode case
+	op := vm.StringToOp(t.config.OpCode)
+	if op == 0 && t.config.OpCode != "STOP" && t.config.OpCode != "" {
+		t.config.OpCode = "inv"
+	}
+	return t, nil
+}
+
+func (t *contractTracer) CaptureStart(env *vm.EVM, from common.Address, to common.Address, create bool, input []byte, gas uint64, value *big.Int) {
+	if create {
+		validateAndStoreOpCode(t, input, to)
+	}
+}
+
+func (t *contractTracer) CaptureEnd(output []byte, gasUsed uint64, _ time.Duration, err error) {
+}
+
+func (t *contractTracer) CaptureState(env *vm.EVM, pc uint64, op vm.OpCode, gas, cost uint64, scope *vm.ScopeContext, rData []byte, depth int, err error) {
+}
+
+func (t *contractTracer) CaptureFault(env *vm.EVM, pc uint64, op vm.OpCode, gas, cost uint64, _ *vm.ScopeContext, depth int, err error) {
+}
+
+func (t *contractTracer) CaptureEnter(typ vm.OpCode, from common.Address, to common.Address, input []byte, gas uint64, value *big.Int) {
+	// Skip if tracing was interrupted
+	if atomic.LoadUint32(&t.interrupt) > 0 {
+		// TODO: env.Cancel()
+		return
+	}
+	if typ == vm.CREATE || typ == vm.CREATE2 {
+		validateAndStoreOpCode(t, input, to)
+	}
+}
+
+func (t *contractTracer) CaptureExit(output []byte, gasUsed uint64, err error) {
+}
+
+func (t *contractTracer) GetResult() (json.RawMessage, error) {
+	res, err := json.Marshal(t.Addrs)
+	if err != nil {
+		return nil, err
+	}
+	return json.RawMessage(res), t.reason
+}
+
+func (t *contractTracer) Stop(err error) {
+	t.reason = err
+	atomic.StoreUint32(&t.interrupt, 1)
+}
+
+func validateAndStoreOpCode(t *contractTracer, input []byte, to common.Address) {
+	// If the OpCode is "inv" or if the OpCode is not empty and doesn't match the input, exit early.
+	if t.config.OpCode == "inv" || (t.config.OpCode != "" && !findOpcodes(input, vm.StringToOp(t.config.OpCode))) {
+		return
+	}
+	// If WithByteCode is true, store the input in the address mapping as hex.
+	if t.config.WithByteCode {
+		t.Addrs[addrToHex(to)] = bytesToHex(input)
```

Comment:
This is not quite right. The inout here will not be the actual bytecode. Rather the contract constructor code.